File: src/services/candidate_functions.py
# src/services/candidateFunctions.py
import sqlite3
import re
import os

def getCandidateIdByName(candidate_name):
    """
    Get a candidate's ID from the database given their name.
    
    This function implements multiple search strategies to handle different name formats:
    1. Exact match in both candidates_master and candidates tables
    2. Name format conversion (First Last <-> Last, First)
    3. Last name priority matching
    4. Partial matching with individual terms
    
    Args:
        candidate_name (str): The name of the candidate to search for.
                             Can be in formats like:
                             - "Thomas Cotton"
                             - "COTTON, THOMAS" 
                             - "Cotton"
                             
    Returns:
        str or None: The candidate ID if found, None otherwise
        
    Examples:
        >>> getCandidateIdByName('Thomas Cotton')
        'H2AR04083'
        >>> getCandidateIdByName('COTTON, THOMAS')
        'H2AR04083'
        >>> getCandidateIdByName('Cotton')
        'H4CA47127'
    """
    try:
        # Updated path to external SSD location
        db_path = '/Volumes/Extreme SSD/OpenPockets/politicaldata.db'
        
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        
        # Clean and normalize the input name
        candidate_name = candidate_name.strip()
        
        # Try multiple search strategies
        result = None
        
        # Strategy 1: Exact match in both tables
        for table in ['candidates_master', 'candidates']:
            cursor.execute(f'SELECT CAND_ID FROM {table} WHERE CAND_NAME = ? COLLATE NOCASE', (candidate_name,))
            result = cursor.fetchone()
            if result:
                break
        
        # Strategy 2: Try reversed name format (First Last -> Last, First)
        if not result and ' ' in candidate_name:
            name_parts = candidate_name.split()
            if len(name_parts) >= 2:
                # Convert "Thomas Cotton" to "COTTON, THOMAS"
                reversed_name = f"{name_parts[-1]}, {' '.join(name_parts[:-1])}"
                for table in ['candidates_master', 'candidates']:
                    cursor.execute(f'SELECT CAND_ID FROM {table} WHERE CAND_NAME = ? COLLATE NOCASE', (reversed_name,))
                    result = cursor.fetchone()
                    if result:
                        break
        
        # Strategy 3: Try original format (Last, First -> First Last)
        if not result and ',' in candidate_name:
            name_parts = candidate_name.split(',')
            if len(name_parts) >= 2:
                # Convert "COTTON, THOMAS" to "Thomas Cotton"
                first_name = name_parts[1].strip()
                last_name = name_parts[0].strip()
                normal_name = f"{first_name} {last_name}"
                for table in ['candidates_master', 'candidates']:
                    cursor.execute(f'SELECT CAND_ID FROM {table} WHERE CAND_NAME = ? COLLATE NOCASE', (normal_name,))
                    result = cursor.fetchone()
                    if result:
                        break
        
        # Strategy 4: Partial match with last name priority
        if not result:
            # Extract potential last name (last word if space-separated, first word if comma-separated)
            if ',' in candidate_name:
                last_name = candidate_name.split(',')[0].strip()
            else:
                last_name = candidate_name.split()[-1] if ' ' in candidate_name else candidate_name
            
            # Search for last name match first (more specific)
            for table in ['candidates_master', 'candidates']:
                cursor.execute(f'SELECT CAND_ID, CAND_NAME FROM {table} WHERE CAND_NAME LIKE ? COLLATE NOCASE ORDER BY CAND_NAME LIMIT 1', 
                              (f'{last_name},%',))
                result = cursor.fetchone()
                if result:
                    break
        
        # Strategy 5: General partial match
        if not result:
            # Try partial match with any part of the name
            search_terms = re.split(r'[,\s]+', candidate_name)
            search_terms = [term.strip() for term in search_terms if term.strip()]
            
            for term in search_terms:
                if len(term) >= 3:  # Only search for terms with 3+ characters
                    for table in ['candidates_master', 'candidates']:
                        cursor.execute(f'SELECT CAND_ID, CAND_NAME FROM {table} WHERE CAND_NAME LIKE ? COLLATE NOCASE ORDER BY CAND_NAME LIMIT 1', 
                                      (f'%{term}%',))
                        result = cursor.fetchone()
                        if result:
                            break
                    if result:
                        break
            
        conn.close()
        return result[0] if result else None
    except Exception as e:
        print(f"Error finding candidate ID: {e}")
        return None

# Example usage and testing (uncomment to test)
if __name__ == "__main__":
    # Test cases
    test_cases = [
        'Thomas Cotton',
        'Cotton', 
        'Thomas',
        'COTTON, THOMAS',
        'Steven Cotton'
    ]
    
    print("Testing getCandidateIdByName function:")
    for test_name in test_cases:
        result = getCandidateIdByName(test_name)
        print(f"  '{test_name}' -> {result}")

File: src/services/candidateFunctions.ts
// TypeScript wrapper for Python candidate functions
import { open } from 'sqlite';
import sqlite3 from 'sqlite3';
import path from 'path';
import { fileURLToPath } from 'url';

// reduce api calls pls
const nameCache: Record<string, string> = {};

export async function getCandidateNameFromDb(candidateId: string): Promise<string | null> {
  try {
    
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    const dbPath = path.resolve(__dirname, 'politicaldata.db');
    
    // is this efficient 
    const db = await open({
      filename: dbPath,
      driver: sqlite3.Database
    });
    
    
    const candidate = await db.get(
      'SELECT CAND_NAME FROM candidates_master WHERE CAND_ID = ?',
      candidateId
    );
    
    if (candidate && candidate.CAND_NAME) {
      await db.close();
      return candidate.CAND_NAME;
    }
    
    // Then try senate table
    const senator = await db.get(
      'SELECT name FROM senate WHERE id = ?',
      candidateId
    );
    
    await db.close();
    
    if (senator && senator.name) {
      return senator.name;
    }
    
    return null;
  } catch (error) {
    console.error(`Error accessing database for candidate ${candidateId}:`, error);
    return null;
  }
}

/**
 * Get a candidate's name from the API given their ID
 */
export const getCandidateName = async (nameOrId: string): Promise<string> => {
  // If it's already a name (not an ID format), return it
  if (!nameOrId.match(/^[A-Z][0-9][A-Z]{2}[0-9]{5}$/) && 
      !nameOrId.match(/Senator \(ID: ([A-Z0-9]+)\)/)) {
    return nameOrId;
  }
  
  // Extract ID if in the format "Senator (ID: XXX)"
  let candidateId = nameOrId;
  const idMatch = nameOrId.match(/Senator \(ID: ([A-Z0-9]+)\)/);
  if (idMatch && idMatch[1]) {
    candidateId = idMatch[1];
  }
  
  // Check cache first
  if (nameCache[candidateId]) {
    return nameCache[candidateId];
  }
  
  // First check local database
  try {
    const nameFromDb = await getCandidateNameFromDb(candidateId);
    if (nameFromDb) {
      // Cache the result
      nameCache[candidateId] = nameFromDb;
      return nameFromDb;
    }
  } catch (dbError) {
    console.error(`Database error for ${candidateId}:`, dbError);
    // Continue to API if database fails
  }
  
  // Then try API if not found in database
  try {
    // Call our API endpoint that uses the Python function
    const response = await fetch(`http://localhost:3001/api/candidate-info/${candidateId}`);
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    const data = await response.json();
    
    if (data && data.name) {
      // Cache the result
      nameCache[candidateId] = data.name;
      return data.name;
    }
  } catch (error) {
    console.error(`Error fetching name for ${candidateId}:`, error);
  }
  
  // Fallback to original string if all methods fail
  return nameOrId;
} 

/**
 * Get a candidate's ID from the database given their name
 * @param candidateName The full name of the candidate to search for
 * @returns The candidate ID if found, null otherwise
 */
export async function getCandidateIdByName(candidateName: string): Promise<string | null> {
    const sqlite3 = require('sqlite3');
    const { open } = require('sqlite');
    const path = require('path');
    
    // Open database connection
    const db = await open({
      filename: path.resolve(__dirname, 'politicaldata.db'),
      driver: sqlite3.Database
    });
    
    try {
      // query for candidate with exact name match
      const candidate = await db.get(
        'SELECT CAND_ID FROM candidates_master WHERE CAND_NAME = ? COLLATE NOCASE',
        [candidateName]
      );
      
      // if no exact match, try partial match
      if (!candidate) {
        const fuzzyMatch = await db.get(
          'SELECT CAND_ID FROM candidates_master WHERE CAND_NAME LIKE ? COLLATE NOCASE LIMIT 1',
          [`%${candidateName}%`]
        );
        return fuzzyMatch ? fuzzyMatch.CAND_ID : null;
      }
      
      return candidate.CAND_ID;
    } catch (error) {
      console.error(`Error finding candidate ID for "${candidateName}":`, error);
      return null;
    } finally {
      await db.close();
    }
  }